docs/architecture/adr-050-sign-mode-textual.md

```diff
@@ -23,72 +24,119 @@ Protobuf-based SIGN_MODE_DIRECT was introduced in [ADR-020](./adr-020-protobuf-t
-- by using property testing on the proto transaction itself, and testing that the composition of encoding and decoding yields the original transaction itself.
+Note that the existence of an inverse function ensures that the
+rendered text contains the full information of the original transaction,
+not a hash or subset.
 
-This also prevents users signing over any hashed transaction data (fee, transaction body, `Msg` content that might be hashed etc), which is a security concern for Ledger's security team.
+### Chain State
 
-We propose to maintain functional tests using bijectivity in the SDK.
+The rendering function (and parsing function) may depend on the current chain state.
+This is useful for reading parameters, such as coin display metadata,
+or for reading user-specific preferences such as language or address aliases.
+Note that if the observed state changes between signature generation
+and the transaction's inclusion in a block, the delivery-time rendering
+might differ. If so, the signature will be invalid and the transaction
+will be rejected.
 
-### 2. Only ASCII 32-127 characters allowed
+### Signature and Security
 
-Ledger devices have limited character display capabilities, so all strings MUST only contain ASCII characters in the 32-127 range.
+For security, transaction signatures should have three properties:
 
-In particular, the newline `"\n"` (ASCII: 10) character is forbidden.
+1. Given the transaction, signatures, and chain state, it must be possible to validate that the signatures matches the transaction,
+to verify that the signers must have known their respective secret keys.
 
-### 3. Strings SHOULD have the `<key>: <value>` format
+2. It must be computationally infeasible to find a substantially different transaction for which the given signatures are valid, given the same chain state.
 
-Given the Regex `/^(\* )?(>* )?(.*: )?(.*)$/`, all strings SHOULD match the Regex with capture groups 3 and 4 non-empty. This is helpful for UIs displaying SignDocTextual to users.
+3. The user should be able to give informed consent to the signed data via a simple, secure device with limited display capabilities.
 
-- The initial `*` character is optional and denotes the Ledger Expert mode, see #5.
-- Strings can also include a number of `>` character to denote nesting.
-- In the case where the first Regex capture group is not empty, it represents an indicative key, whose associated value is given in the second capture group. This MAY be used in the Ledger app to perform custom on-screen formatting, for example to break long lines into multiple screens.
+The correctness and security of `SIGN_MODE_TEXTUAL` is guaranteed by demonstrating an inverse function from the rendering to transaction protos.
+This means that it is impossible for a different protocol buffer message to render to the same text.
```

Comment:
It appears that inversability is not guaranteed or even verified by the SDK, right? It's just assumed to be true. So if an implementation happens to have a bug that violates this assumption, and isn't fully inversible in some circumstance, what happens?